server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationStatsService.java

```diff
@@ -18,78 +18,102 @@
+    private final StatsProvider statsProvider;
 
     public AllocationStatsService(
         ClusterService clusterService,
         ClusterInfoService clusterInfoService,
         ShardsAllocator shardsAllocator,
         WriteLoadForecaster writeLoadForecaster
     ) {
-        this.clusterService = clusterService;
-        this.clusterInfoService = clusterInfoService;
-        this.desiredBalanceShardsAllocator = shardsAllocator instanceof DesiredBalanceShardsAllocator allocator ? allocator : null;
-        this.writeLoadForecaster = writeLoadForecaster;
+        this.statsProvider = new NodeStatsProvider(clusterService, clusterInfoService, writeLoadForecaster);
+        this.desiredBalanceSupplier = shardsAllocator instanceof DesiredBalanceShardsAllocator allocator
+            ? allocator::getDesiredBalance
+            : () -> null;
     }
 
     public Map<String, NodeAllocationStats> stats() {
-        var state = clusterService.state();
-        var info = clusterInfoService.getClusterInfo();
-        var desiredBalance = desiredBalanceShardsAllocator != null ? desiredBalanceShardsAllocator.getDesiredBalance() : null;
+        return statsProvider.stats(desiredBalanceSupplier.get());
+    }
 
-        var stats = Maps.<String, NodeAllocationStats>newMapWithExpectedSize(state.getRoutingNodes().size());
-        for (RoutingNode node : state.getRoutingNodes()) {
-            int shards = 0;
-            int undesiredShards = 0;
-            double forecastedWriteLoad = 0.0;
-            long forecastedDiskUsage = 0;
-            long currentDiskUsage = 0;
-            for (ShardRouting shardRouting : node) {
-                if (shardRouting.relocating()) {
-                    continue;
-                }
-                shards++;
-                IndexMetadata indexMetadata = state.metadata().getIndexSafe(shardRouting.index());
-                if (isDesiredAllocation(desiredBalance, shardRouting) == false) {
-                    undesiredShards++;
-                }
-                long shardSize = info.getShardSize(shardRouting.shardId(), shardRouting.primary(), 0);
-                forecastedWriteLoad += writeLoadForecaster.getForecastedWriteLoad(indexMetadata).orElse(0.0);
-                forecastedDiskUsage += Math.max(indexMetadata.getForecastedShardSizeInBytes().orElse(0), shardSize);
-                currentDiskUsage += shardSize;
+    @FunctionalInterface
+    public interface StatsProvider {
+        Map<String, NodeAllocationStats> stats(DesiredBalance desiredBalance);
+    }
 
-            }
-            stats.put(
-                node.nodeId(),
-                new NodeAllocationStats(
-                    shards,
-                    desiredBalanceShardsAllocator != null ? undesiredShards : -1,
-                    forecastedWriteLoad,
-                    forecastedDiskUsage,
-                    currentDiskUsage
-                )
-            );
+    public static class NodeStatsProvider implements StatsProvider {
```

Comment:
I find it a bit confusing with the inner class, could we just add an overload to `AllocationStatsService` that took a desired balance to use in the computation, and everyone just gets given a reference to a single instance of it?

e.g. add
```java
public Map<String, NodeAllocationStats> stats(DesiredBalance desiredBalance) {
   // ...
}
```

Reply:
Ahh there would be a circular dependency between allocator and stats service then.

I reckon it'd be clearer if the thing that calculated the stats just became a separate shared piece of infrastructure that took a `DesiredBalance` from somewhere and returned the stats. The inner class and the interface makes it a bit tricky to read.

Reply:
Yes, the dependencies here is not straight-forward and largely the reason I came up with this. But how is your suggestion different than what's in the PR? I can move out the inner class and get rid of the interface. Other than that, it is the same, no?

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/116585/commits/a5381e1c9d8913b7ca8273e41164bc337f23beff.